### PR TITLE
refactor: rewrite perl provider in lua

### PR DESCRIPTION
**Pull request by dundargoc**

### CODE REVIEW

File: runtime/lua/vim/provider/perl.lua [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+--- @return string? error message if perl can't be detected; nil if success
+function M.detect()
+  -- use g:perl_host_prog if set or check if perl is on the path
+  local prog = vim.fn.exepath(vim.g.perl_host_prog or 'perl')
+  if prog == '' then
+    return nil, 'No perl executable found'
+  end
+
+  -- if perl is available, make sure we have 5.22+
+  vim.fn.system({ prog, '-e', 'use v5.22' })
```
Comment (dundargoc) [inline runtime/lua/vim/provider/perl.lua:28]: `vim.system` does not work here, I think there's a bug.

Comment (wookayin) [reply runtime/lua/vim/provider/perl.lua:28]: interesting. what happens exactly?

Comment (dundargoc) [reply runtime/lua/vim/provider/perl.lua:28]: `vim.system` just silently aborts in one of the tests, causing it to fail.

File: runtime/lua/provider/perl/health.lua [hunk 2]
```diff
@@ -9,19 +9,14 @@ function M.check()
...
-      })
-    else
-      health.warn('No usable perl executable found')
-    end
+  if not perl_exec then
+    health.warn(assert(perl_warnings), {
+      'See :help provider-perl for more information.',
+      'You may disable this provider (and warning) by adding `let g:loaded_perl_provider = 0` to your init.vim',
+    })
+    health.warn('No usable perl executable found')
```
Comment (wookayin) [inline runtime/lua/provider/perl/health.lua:19]: When user set an invalid `g:perl_host_prog`, it doesn't speak anything about the wrong `g:perl_host_prog`. Like in python provider helathcheck, it'd be good to print out `g:perl_host_prog` if set.

**Review by wookayin — APPROVED** [review]

File: runtime/autoload/provider/perl.vim [hunk 3]
```diff
@@ -2,82 +2,14 @@ if exists('s:loaded_perl_provider')
   finish
 endif
 
-let s:loaded_perl_provider = 1
```
Comment (wookayin) [inline runtime/autoload/provider/perl.vim:5]: Now that `s:loaded_perl_provider = 1` is now gone, Line 1 should be changed to:
```diff
-if exists('s:loaded_perl_provider')
+if exists('g:loaded_perl_provider')
```

File: runtime/lua/provider/perl/health.lua (hunk 2 above)
Comment (dundargoc) [reply runtime/lua/provider/perl/health.lua:19]: Yeah, I think I'd prefer to just dump the code over to the lua side as is and treat any logical improvements as out of scope for the "lua-ify" PRs.

**Pull request merged**